### PR TITLE
[Merged by Bors] - feat: port Algebra.Order.Ring.WithTop

### PR DESCRIPTION
**Pull request by mcdoll**

### CODE REVIEW

**Comment by mcdoll** [comment]
Needs either the `lift` tactic or some nontrivial rewriting

**Comment by LukasMias** [comment]
I think this file should be ready now, but one needs to be mindful of the porting note
```lean
-- Todo (porting): Move this
instance : DecidableEq (WithTop α) := instDecidableEqOption
instance : DecidableEq (WithBot α) := instDecidableEqOption
```
placed by @mcdoll. I can't comment on that note though, because I don't know where to ```Move this``` :)

**Comment by mcdoll** [comment]
Still on vacation, but I think these instances should be closer to the definition of `WithTop α` and `WithBot α`. It seems that in Lean 3, they were derived automatically, but in Lean 4 this is no longer the case.

File: Mathlib/Algebra/Order/Ring/WithTop.lean [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+-/
+
+section foo
+
+variable [DecidableEq α]
+
+end foo
+
+set_option autoImplicit false
+
```
Comment (dupuisf) [inline Mathlib/Algebra/Order/Ring/WithTop.lean:27]:
```suggestion
```

**Review by dupuisf — COMMENTED** [review]
Looks good!

bors d+

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**